h2o-core/src/main/java/water/rapids/ast/prims/models/AstMakeLeaderboard.java

```diff
@@ -0,0 +1,184 @@
+import hex.leaderboard.*;
+import water.DKV;
+import water.Key;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.logging.Logger;
+import water.logging.LoggerFactory;
+import water.rapids.Env;
+import water.rapids.ast.AstPrimitive;
+import water.rapids.ast.AstRoot;
+import water.rapids.vals.ValFrame;
+import water.util.TwoDimTable;
+import water.util.Log;
+
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+/**
+ * Compute Leaderboard
+ */
+public class AstMakeLeaderboard extends AstPrimitive {
+
+    @Override
+    public String[] args() {
+        return new String[]{"models", "leaderboardFrame", "sortMetric", "extensions", "scoringData"};
+    }
+
+    @Override
+    public int nargs() {
+        return 1 + 5;
+    } // (makeLeaderboard models leaderboardFrame sortMetric extensions projectName)
+
+    @Override
+    public String str() {
+        return "makeLeaderboard";
+    }
+
+    private static LeaderboardExtensionsProvider createLeaderboardExtensionProvider(Frame leaderboardFrame) {
+
+        return new LeaderboardExtensionsProvider() {
+            @Override
+            public LeaderboardCell[] createExtensions(Model model) {
+                return new LeaderboardCell[]{
+                        new TrainingTime(model),
+                        new ScoringTimePerRow(model, leaderboardFrame),
+                        new AlgoName(model),
+                };
+            }
+        };
+    }
+
+    @Override
+    public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
+        Key[] models = Arrays.stream(stk.track(asts[1].exec(env)).getStrs())
+                .flatMap(model_id -> {
+                    Object obj = DKV.getGet(model_id);
+                    if (obj instanceof Model) {
+                        return Stream.of(Key.make(model_id));
+                    } else if (obj instanceof ModelContainer) {
+                        ModelContainer mc = (ModelContainer) obj;
+                        return Stream.of(mc.getModelKeys());
+                    } else {
+                        throw new RuntimeException("Unsupported model/grid id: " + model_id + "!");
+                    }
+                }).toArray(Key[]::new);
+
+        // Get Frame clones Frame and the clone is key-less and since we are doing just read only ops here we can use the original frame
+        String leaderboardFrameKey = stk.track(asts[2].exec(env)).getStr();
+        Frame leaderboardFrame = null;
+        if (!leaderboardFrameKey.isEmpty())
+            leaderboardFrame = DKV.getGet(leaderboardFrameKey);
+        String sortMetric = stk.track(asts[3].exec(env)).getStr();
+        String[] extensions = stk.track(asts[4].exec(env)).getStrs();
+        String scoringData = stk.track(asts[5].exec(env)).getStr().toLowerCase();
+
+        String projectName = leaderboardFrameKey + "_" + Arrays.deepHashCode(models) + "_" + sortMetric + "_" +
+                Arrays.deepHashCode(extensions) + "_" + scoringData;
```

Comment:
not sure why the hashed extensions should be part of the projectName...
The same leaderboard can be used with a different set of extensions, even if the `extensionProvider` is set later as extensions should be able to generate their data by themselves.